app/controllers/concerns/application_controller/crud.rb

```diff
@@ -45,30 +45,46 @@ def set_success_and_redirect(obj, options = {})
   end
 
   # gets the url to an index action, ensuring the appropriate page is returned to
-  # ctlr - the controller whose index should be used. defaults to current controller
-  def index_url_with_page_num(ctlr = nil)
-    url_for(:controller => ctlr || controller_name, :action => :index, :page => get_last_page_number)
+  # target_controller - the controller whose index should be used. defaults to current controller
+  def index_url_with_context(target_controller = nil)
+    target_controller ||= controller_name
+    url_params = {controller: target_controller, action: :index}.merge(get_last_context)
+    url_for(url_params)
   end
 
   # remembers the last visited page number for each controller and mission
-  def remember_page_number
-    if params[:page]
-      session[:last_page_numbers] ||= {}
-      session[:last_page_numbers][last_page_number_hash_key] = params[:page]
-    end
+  def remember_context
+    remember_search(params[:search]) if params[:search]
+    remember_page(params[:page]) if params[:page]
+  end
+
+  def remember_search(search)
+    session[:last_search] ||= {}
+    session[:last_search][last_context_hash_key] = search
+  end
+
+  def remember_page(page)
+    session[:last_page_numbers] ||= {}
+    session[:last_page_numbers][last_context_hash_key] = page
   end
 
-  # builds a simple hash key for remembering page numbers
-  def last_page_number_hash_key
-    controller_name + current_mission.try(:id).to_s
+  # builds a simple hash key for remembering page context
+  def last_context_hash_key
+    "#{controller_name}#{current_mission.try(:id)}"
+  end
+
+  def get_last_context
+    page = get_last_page_number if session[:last_page_numbers]
+    search = get_last_search if session[:last_search]
+    { page: page, search: search }
   end
 
   def get_last_page_number
-    if session[:last_page_numbers]
-      session[:last_page_numbers][last_page_number_hash_key]
-    else
-      nil
-    end
+    session[:last_page_numbers][last_context_hash_key].presence
+  end
+
+  def get_last_search
+    session[:last_search][last_context_hash_key].presence
```

Comment:
Using `.presence` here can still end up causing a nil error if `session[:last_search]` is nil. You could do `session[:last_search].try(:[], last_context_hash_key)` though?

Also above.